device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+
+  private _mqtt: MqttBase;
+  private _pendingTwinRequests: { [key: string]: any } = {};
+  private _topicFsm: machina.BehavioralFsm;
+  private _responseTopic: TopicSubscription;
+  private _desiredPropertiesUpdatesTopic: TopicSubscription;
+
+  constructor(client: MqttBase) {
+    super();
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [The `MqttTwinClient` constructor shall accept a `client` object.]*/
+    this._mqtt = client;
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object.]*/
+    this._mqtt.on('message', messageHandler);
+
+    this._topicFsm = new machina.BehavioralFsm({
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: function (client: TopicSubscription, err: Error, callback: (err?: Error) => void): void {
+            if (callback) {
+              callback(err);
+            }
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'subscribing', callback);
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            // not entirely sure about that. if subscription are restored because cleanSession is false, it means technically a user may want to unsubscribe
+            // even though subscribe hasn't been called yet.
+            callback();
+          }
+        },
+        subscribing: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            client.mqttClient.subscribe(client.topic, { qos: 0 }, (err, result) => {
+              if (err) {
+                this.transition(client, 'unsubscribed', err, callback);
+              } else {
+                debug('subscribed to response topic: ' + JSON.stringify(result));
+                this.transition(client, 'subscribed', callback);
+              }
+            });
+          },
+          '*': function (client: TopicSubscription): void { this.deferUntilTransition(client); }
+        },
+        subscribed: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'unsubscribing', callback);
+          }
+        },
+        unsubscribing: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            client.mqttClient.unsubscribe(client.topic, (err, result) => {
```

Comment:
>unsubscribe [](start = 30, length = 11)

See comment above on unsubscribing.  It may also be that this comment applies only to the PATCH topic and not to the RES topic, but it's so goofy that I can't be confident about anything I'm saying.  IIRC, Max is your man.

Reply:
I will double check.

---
In reply to: [170699475](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170699475) [](ancestors = 170699475)

Reply:
according to Robledo we should unsubscribe.

---
In reply to: [171443628](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r171443628) [](ancestors = 171443628,170699475)